File: go/tooling/hosted.go
package tooling

import (
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/bettercorp/service-base/go/bsb"
)

type HostedOptions struct {
	Plugin, Language, Version string
	AllowHTTP                 bool
}

func hostedOrigin(endpoint string, allowHTTP bool) (string, error) {
	origin, err := bsb.EndpointOrigin(endpoint, allowHTTP)
	if err != nil {
		return "", err
	}
	base, _ := url.Parse(origin)
	if base.Port() == "443" && base.Scheme == "https" || base.Port() == "80" && base.Scheme == "http" {
		base.Host = base.Hostname()
		if strings.Contains(base.Host, ":") {
			base.Host = "[" + base.Host + "]"
		}
		origin = base.String()
	}
	return origin, nil
}
func HostedSchema(ctx context.Context, endpoint string, options HostedOptions) (map[string]any, string, error) {
	origin, err := hostedOrigin(endpoint, options.AllowHTTP)
	if err != nil {
		return nil, "", err
	}
	discovery, _ := url.Parse(origin + "/.well-known/bsb")
	fetch := func(endpoint string) (map[string]any, error) {
		result := map[string]any{}
		err := bsb.JSONRequest(ctx, http.MethodGet, endpoint, nil, nil, 10*time.Second, &result)
		return result, err
	}
	manifest, err := fetch(discovery.String())
	if err != nil {
		return nil, "", err
	}
	revision, ok := manifest["bsb"].(json.Number)
	revisionNumber, revisionErr := revision.Float64()
	if !ok || revisionErr != nil || revisionNumber != 1 {
		return nil, "", fmt.Errorf("invalid BSB discovery version")
	}
	entries, ok := manifest["plugins"].([]any)
	if !ok || len(entries) == 0 || len(entries) > 128 {
		return nil, "", fmt.Errorf("invalid hosted plugins")
	}
	selectedOrg, selectedName := "", ""
	if options.Plugin != "" {
		selectedOrg, selectedName, err = ParsePluginID(options.Plugin)
		if err != nil {
			return nil, "", err
		}
	}
	if options.Language != "" {
		options.Language, err = language(options.Language)
		if err != nil {
			return nil, "", err
		}
	}
	if options.Version != "" && !versionPattern.MatchString(options.Version) {
		return nil, "", fmt.Errorf("exact semantic version required")
	}
	identities := map[string]bool{}
	matches := []map[string]any{}
	for _, raw := range entries {
		entry, ok := raw.(map[string]any)
		if !ok {
			return nil, "", fmt.Errorf("invalid hosted plugin")
		}
		id, _ := entry["id"].(string)
		org, name, err := ParsePluginID(id)
		if err != nil {
			return nil, "", err
		}
		impl, _ := entry["language"].(string)
		impl, err = language(impl)
		if err != nil {
			return nil, "", err
		}
		version, _ := entry["version"].(string)
		if !versionPattern.MatchString(version) {
			return nil, "", fmt.Errorf("invalid hosted version")
		}
		switch entry["schema"].(type) {
		case string, map[string]any:
		default:
			return nil, "", fmt.Errorf("invalid hosted schema reference")
		}
		identity := org + "/" + name + "~" + impl + "~" + version
		if identities[identity] {
			return nil, "", fmt.Errorf("duplicate hosted implementation")
		}
		identities[identity] = true
		if (options.Plugin == "" || (org == selectedOrg && name == selectedName)) && (options.Language == "" || impl == options.Language) && (options.Version == "" || version == options.Version) {
			matches = append(matches, map[string]any{"org": org, "name": name, "language": impl, "version": version, "schema": entry["schema"]})
		}
	}
	if len(matches) != 1 {
		return nil, "", fmt.Errorf("no unique hosted implementation; specify --plugin, --source-language or --version")
	}
	selected := matches[0]
	schema, inline := selected["schema"].(map[string]any)
	if !inline {
		ref, err := url.Parse(selected["schema"].(string))
		if err != nil {
			return nil, "", err
		}
		link := discovery.ResolveReference(ref)
		authority := *link
		authority.Path = ""
		authority.RawPath = ""
		authority.RawQuery = ""
		authority.Fragment = ""
		linkedOrigin, err := hostedOrigin(authority.String(), options.AllowHTTP)
		if err != nil || linkedOrigin != origin || link.User != nil || link.Fragment != "" {
			return nil, "", fmt.Errorf("hosted schema link must remain on the same origin without credentials or fragments")
		}
		schema, err = fetch(link.String())
		if err != nil {
			return nil, "", err
		}
	}
	if _, ok := schema["events"].(map[string]any); !ok {
		return nil, "", fmt.Errorf("hosted schema requires an events object")
	}
	target, exists := schema["pluginId"]
	if !exists {
		target, exists = schema["pluginName"]
	}
	if !exists {
		target = selected["name"]
	}
	wire, ok := target.(string)
	_, name, err := ParsePluginID(wire)
	if !ok || err != nil || name != wire {
		return nil, "", fmt.Errorf("invalid hosted wire plugin ID")
	}
	schema["pluginId"] = wire
	if version, exists := schema["version"]; exists && version != selected["version"] {
		return nil, "", fmt.Errorf("hosted schema version does not match discovery")
	}
	schema["version"] = selected["version"]
	schema["source"] = map[string]any{"url": origin, "org": selected["org"], "name": selected["name"], "language": selected["language"], "version": selected["version"]}
	digest := sha256.Sum256([]byte(origin))
	local := fmt.Sprintf("hosted~%x~%s~%s~%s", digest[:8], selected["org"], selected["name"], selected["language"])
	return schema, local, nil
}

func InstallHosted(ctx context.Context, cwd, endpoint string, options HostedOptions) ([]string, error) {
	schema, local, err := HostedSchema(ctx, endpoint, options)
	if err != nil {
		return nil, err
	}
	return installSchema(cwd, schema, local)
}

File: rust/cli/src/hosted.rs
use crate::tooling::{exact_version, install_schema, plugin_id, source_language};
use anyhow::{Context, Result, ensure};
use bsb::{config::MAX_JSON, http};
use reqwest::{Method, Url, header::HeaderMap};
use serde::Deserialize;
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use std::{collections::BTreeSet, path::Path, time::Duration};

#[derive(Deserialize)]
struct Discovery {
    bsb: u64,
    plugins: Vec<Entry>,
}

#[derive(Deserialize)]
struct Entry {
    id: String,
    language: String,
    version: String,
    schema: Value,
}

pub struct HostedClient {
    origin: Url,
    canonical_origin: String,
}

impl HostedClient {
    pub fn new(origin: &str, allow_http: bool) -> Result<Self> {
        let mut origin = http::origin(origin, allow_http)?;
        if let Some(host) = origin.host_str() {
            origin.set_host(Some(&host.to_ascii_lowercase()))?;
        }
        if matches!(
            (origin.scheme(), origin.port()),
            ("http", Some(80)) | ("https", Some(443))
        ) {
            ensure!(origin.set_port(None).is_ok(), "invalid hosted origin port");
        }
        let canonical_origin = origin.as_str().trim_end_matches('/').to_owned();
        Ok(Self {
            origin,
            canonical_origin,
        })
    }

    async fn request(&self, url: Url) -> Result<Value> {
        http::request(
            &http::client(Duration::from_secs(10))?,
            Method::GET,
            url,
            HeaderMap::new(),
            None,
            MAX_JSON,
        )
        .await
    }

    fn valid_entry(entry: &Entry) -> Result<(String, String, String)> {
        let (org, name) = plugin_id(&entry.id)?;
        let language = source_language(&entry.language)?;
        ensure!(
            exact_version(&entry.version),
            "exact semantic version required"
        );
        Ok((org, name, language))
    }

    fn schema_url(&self, link: &str) -> Result<Url> {
        let url = self.origin.join(".well-known/bsb")?.join(link)?;
        ensure!(
            url.origin() == self.origin.origin()
                && url.username().is_empty()
                && url.password().is_none()
                && url.fragment().is_none(),
            "schema link must remain on the hosted origin"
        );
        Ok(url)
    }

    pub async fn install(
        &self,
        cwd: &Path,
        selected: Option<&str>,
        source: Option<&str>,
        version: Option<&str>,
    ) -> Result<Vec<std::path::PathBuf>> {
        let discovery: Discovery =
            serde_json::from_value(self.request(self.origin.join(".well-known/bsb")?).await?)?;
        ensure!(discovery.bsb == 1, "unsupported hosted discovery version");
        ensure!(discovery.plugins.len() <= 128, "too many hosted plugins");
        let selected = selected.map(plugin_id).transpose()?;
        let source = source.map(source_language).transpose()?;
        if let Some(version) = version {
            ensure!(exact_version(version), "exact semantic version required");
        }
        let mut identities = BTreeSet::new();
        let mut matches = Vec::new();
        for entry in &discovery.plugins {
            let (org, name, language) = Self::valid_entry(entry)?;
            ensure!(
                entry.schema.is_object() || entry.schema.is_string(),
                "hosted schema must be an object or link"
            );
            ensure!(
                identities.insert((
                    org.clone(),
                    name.clone(),
                    language.clone(),
                    entry.version.clone(),
                )),
                "duplicate hosted plugin identity"
            );
            if selected
                .as_ref()
                .is_none_or(|(selected_org, selected_name)| {
                    selected_org == &org && selected_name == &name
                })
                && source.as_deref().is_none_or(|value| value == language)
                && version.is_none_or(|value| value == entry.version)
            {
                matches.push((entry, org, name, language));
            }
        }
        ensure!(
            matches.len() == 1,
            "hosted plugin selection is ambiguous or absent"
        );
        let (entry, org, name, language) = matches.pop().unwrap();
        let mut schema = match &entry.schema {
            Value::Object(_) => entry.schema.clone(),
            Value::String(link) => self.request(self.schema_url(link)?).await?,
            _ => anyhow::bail!("hosted schema must be an object or link"),
        };
        ensure!(
            schema.get("events").is_some_and(Value::is_object),
            "hosted schema events required"
        );
        if schema.get("pluginId").is_none() {
            schema["pluginId"] = schema
                .get("pluginName")
                .cloned()
                .unwrap_or_else(|| json!(name));
        }
        let target = schema["pluginId"]
            .as_str()
            .context("hosted schema plugin ID required")?;
        let (_, target_name) = plugin_id(target)?;
        ensure!(
            target_name == target,
            "hosted schema plugin ID must be unqualified"
        );
        if let Some(version) = schema.get("version") {
            ensure!(
                version.as_str() == Some(entry.version.as_str()),
                "hosted schema version must match discovery"
            );
        } else {
            schema["version"] = json!(entry.version);
        }
        schema["source"] = json!({
            "url": self.canonical_origin,
            "org": org,
            "name": name,
            "language": language,
            "version": entry.version,
        });
        let hash = format!("{:x}", Sha256::digest(self.canonical_origin.as_bytes()));
        let local = format!("hosted~{}~{org}~{name}~{language}", &hash[..16]);
        install_schema(cwd, &schema, &local).await
    }
}
